directembedding/src/main/scala/ch/epfl/directembedding/transformers/EmbeddingTransformation.scala

```diff
@@ -4,35 +4,67 @@ import ch.epfl.directembedding.{ TypeHelper, DirectEmbeddingModule, DirectEmbedd
 
 class reifyAs(to: Any) extends scala.annotation.StaticAnnotation
 class ignore extends scala.annotation.StaticAnnotation
+class reifyAsInvoked extends scala.annotation.StaticAnnotation
+class reifyAsInvokedWithMethod(method: String) extends scala.annotation.StaticAnnotation
 
-trait ReifyAsEmbedding extends DirectEmbeddingModule
+trait EmbeddingTransformation extends DirectEmbeddingModule
   with DirectEmbeddingUtils
+  with ReificationTransformation
   with TypeHelper {
   import c.universe._
 
-  object ReifyAsTransformer extends (Tree => Tree) {
+  object EmbeddingTransformer extends (Tree => Tree) {
     def apply(tree: Tree) = {
       logStarred("Before ReifyAsEmbedding")
-      new ReifyAsTransformer().apply(tree)
+      new EmbeddingTransformer().apply(tree)
     }
   }
 
-  final class ReifyAsTransformer extends Transformer {
+  final class EmbeddingTransformer extends Transformer {
     def apply(tree: Tree): Tree = transform(tree)
 
     private def reifyAnnotation(methodSym: Symbol): Option[Annotation] =
       methodSym.annotations.find(_.tree.tpe <:< c.typeOf[reifyAs])
 
-    private def reify(lhs: Tree, tree: Tree, extraTargs: List[Tree], extraArgs: List[Tree]): Tree = {
-      reify(body(lhs, tree), typeArgs(lhs) ::: extraTargs, args(lhs, extraArgs))
+    private def annotations(sym: Symbol, tree: Tree): List[Annotation] = {
+      fallbackAnnotation(sym, tree).toList ::: sym.annotations
     }
 
-    private def reify(body: Tree, targs: List[Tree], args: List[Tree]): Tree = {
-      (targs, args) match {
-        case (Nil, Nil)    => body
-        case (targs, Nil)  => q"$body.apply[..$targs]"
-        case (Nil, args)   => q"$body.apply(..$args)"
-        case (targs, args) => q"$body.apply[..$targs](..$args)"
+    private def reify(lhs: Tree, tree: Tree, extraTargs: List[Tree], extraArgs: List[Tree]): Tree = {
+      val sym = symbol(lhs)
+      val reification: List[Reification] = annotations(sym, tree).collect {
+        case a if a.tree.tpe <:< c.typeOf[reifyAs] =>
+          ReifyAsReification(
+            annotationBody(a),
+            typeArgs(lhs) ::: extraTargs,
+            args(lhs, extraArgs))
+        case a if a.tree.tpe <:< c.typeOf[reifyAsInvoked] =>
+          val s = self(lhs).getOrElse {
+            c.abort(tree.pos, "reifyAsInvoked cannot be annotated to an object method")
```

Comment:
I added a `reifyAsInvokedFrom(object)` annotation, which allows you to specify the self object. I am actually using it for slick-direct.